File: runtime/src/hosts/http.js
/**
 * Performs one HTTP request via the platform `fetch` and serializes the whole response into a single
 * `Uint8Array` for `src/stdlib/net/http_response.dream`: an ASCII head (status line + header lines) and a blank
 * line, then the raw body bytes. Keeping the body raw (an `arrayBuffer`) makes binary responses
 * byte-exact. `body` is either a string or a `Uint8Array` (or "" / empty for none).
 *
 * `httpRequestStream`/`httpRequestStreamBytes` below instead keep the response body's
 * `ReadableStream` reader in a handle table (`streamHandles`) so `httpReadChunk` can pull it
 * incrementally without buffering the whole body — the JS-host mirror of the native host's
 * `reqwest::blocking::Response` handle table (`src/execution/host/http.rs`).
 */
async function httpDo(url, method, headersJson, body, timeoutMs, _httpVersion) {
  const verb = (method || "GET").toUpperCase();
  const init = { method: verb };
  if (headersJson && headersJson !== "") {
    try { init.headers = JSON.parse(headersJson); } catch (_) { /* ignore bad header json */ }
  }
  const hasBody = typeof body === "string" ? body !== "" : body && body.length > 0;
  if (hasBody && verb !== "GET" && verb !== "HEAD") {
    init.body = body;
  }
  const ms = Number(timeoutMs) || 0;
  if (ms > 0) {
    if (typeof AbortSignal !== "undefined" && typeof AbortSignal.timeout === "function") {
      init.signal = AbortSignal.timeout(ms);
    } else {
      const ctrl = new AbortController();
      init.signal = ctrl.signal;
      setTimeout(() => ctrl.abort(), ms);
    }
  }
  try {
    const res = await fetch(url, init);

    let head = `${res.status}\n`;
    res.headers.forEach((value, name) => {
      head += `${name}: ${value}\n`;
    });
    head += "\n"; // blank line separating head from body
    const headBytes = new TextEncoder().encode(head);
    const bodyBytes = new Uint8Array(await res.arrayBuffer());

    const out = new Uint8Array(headBytes.length + bodyBytes.length);
    out.set(headBytes, 0);
    out.set(bodyBytes, headBytes.length);
    return out;
  } catch (e) {
    const msg = (e && (e.name === "TimeoutError" || e.name === "AbortError"))
      ? "timeout"
      : String((e && e.message) || e || "fetch failed");
    const headBytes = new TextEncoder().encode("0\n\n");
    const bodyBytes = new TextEncoder().encode(msg);
    const out = new Uint8Array(headBytes.length + bodyBytes.length);
    out.set(headBytes, 0);
    out.set(bodyBytes, headBytes.length);
    return out;
  }
}

/**
 * Opens a request via `fetch` and returns just the head (status + headers), keeping the response
 * body's `ReadableStream` reader in a handle table for `httpReadChunk` to pull incrementally.
 * Wire format matches [`httpDo`] exactly except the "body" is the decimal handle id, mirroring the
 * native host's `open_http_stream` (`src/execution/host/http.rs`) so `HttpResponse`'s head parser
 * and `HttpStreamResponse` (`wrap_stream` in `http_client.dream`) work unchanged on both hosts.
 */
async function httpDoStream(url, method, headersJson, body, timeoutMs, _httpVersion) {
  const verb = (method || "GET").toUpperCase();
  const init = { method: verb };
  if (headersJson && headersJson !== "") {
    try { init.headers = JSON.parse(headersJson); } catch (_) { /* ignore bad header json */ }
  }
  const hasBody = typeof body === "string" ? body !== "" : body && body.length > 0;
  if (hasBody && verb !== "GET" && verb !== "HEAD") {
    init.body = body;
  }
  const ms = Number(timeoutMs) || 0;
  if (ms > 0) {
    if (typeof AbortSignal !== "undefined" && typeof AbortSignal.timeout === "function") {
      init.signal = AbortSignal.timeout(ms);
    } else {
      const ctrl = new AbortController();
      init.signal = ctrl.signal;
      setTimeout(() => ctrl.abort(), ms);
    }
  }
  try {
    const res = await fetch(url, init);

    let head = `${res.status}\n`;
    res.headers.forEach((value, name) => {
      head += `${name}: ${value}\n`;
    });
    head += "\n";
    const id = nextStreamHandle++;
    streamHandles.set(id, new HttpStreamReader(res.body));
    head += String(id);
    return new TextEncoder().encode(head);
  } catch (e) {
    const msg = (e && (e.name === "TimeoutError" || e.name === "AbortError"))
      ? "timeout"
      : String((e && e.message) || e || "fetch failed");
    const headBytes = new TextEncoder().encode("0\n\n");
    const bodyBytes = new TextEncoder().encode(msg);
    const out = new Uint8Array(headBytes.length + bodyBytes.length);
    out.set(headBytes, 0);
    out.set(bodyBytes, headBytes.length);
    return out;
  }
}

const streamHandles = new Map(); // handle -> HttpStreamReader
let nextStreamHandle = 1;

/** Buffers `ReadableStream` chunks so `httpReadChunk` can hand back exactly `maxBytes` at a time. */
class HttpStreamReader {
  constructor(body) {
    // `fetch` always resolves a body in Node/browsers; a `null` body (e.g. a HEAD response) reads
    // as immediate EOF.
    this.reader = body ? body.getReader() : null;
    this.leftover = null; // Uint8Array not yet handed out
    this.eof = !body;
  }

  async read(maxBytes) {
    const cap = Math.max(1, Number(maxBytes) || 65536);
    if (!this.leftover && !this.eof) {
      const { value, done } = await this.reader.read();
      if (done) {
        this.eof = true;
      } else {
        this.leftover = value;
      }
    }
    if (!this.leftover) {
      return { tag: "eof", bytes: new Uint8Array(0) };
    }
    if (this.leftover.length <= cap) {
      const out = this.leftover;
      this.leftover = null;
      return { tag: "data", bytes: out };
    }
    const out = this.leftover.subarray(0, cap);
    this.leftover = this.leftover.subarray(cap);
    return { tag: "data", bytes: out };
  }

  cancel() {
    if (this.reader) {
      try { this.reader.cancel(); } catch (_) { /* already closed */ }
    }
  }
}

/** Wire: `"data\n<bytes>"` | `"eof\n"` | `"error\n<message>"`, matching `http_read_chunk`. */
async function httpReadChunk(handle, maxBytes) {
  const stream = streamHandles.get(handle);
  if (!stream) {
    return new TextEncoder().encode("error\nstream not found");
  }
  try {
    const { tag, bytes } = await stream.read(maxBytes);
    if (tag === "eof") {
      streamHandles.delete(handle);
    }
    const tagBytes = new TextEncoder().encode(`${tag}\n`);
    const out = new Uint8Array(tagBytes.length + bytes.length);
    out.set(tagBytes, 0);
    out.set(bytes, tagBytes.length);
    return out;
  } catch (e) {
    streamHandles.delete(handle);
    const msg = String((e && e.message) || e || "stream read failed");
    return new TextEncoder().encode(`error\n${msg}`);
  }
}

function httpCloseStream(handle) {
  const stream = streamHandles.get(handle);
  if (!stream) return 0;
  stream.cancel();
  streamHandles.delete(handle);
  return 1;
}

export function makeHttpHost() {
  return {
    // `httpVersion` is accepted for ABI parity with the native host; `fetch` negotiates on its own.
    httpRequest: (url, method, headersJson, body, timeoutMs, httpVersion) =>
      httpDo(url, method, headersJson, body, timeoutMs, httpVersion),
    httpRequestBytes: (url, method, headersJson, body, timeoutMs, httpVersion) =>
      httpDo(url, method, headersJson, Uint8Array.from(body || []), timeoutMs, httpVersion),
    httpRequestStream: (url, method, headersJson, body, timeoutMs, httpVersion) =>
      httpDoStream(url, method, headersJson, body, timeoutMs, httpVersion),
    httpRequestStreamBytes: (url, method, headersJson, body, timeoutMs, httpVersion) =>
      httpDoStream(url, method, headersJson, Uint8Array.from(body || []), timeoutMs, httpVersion),
    httpReadChunk: (handle, maxBytes) => httpReadChunk(handle, maxBytes),
    httpCloseStream: (handle) => httpCloseStream(handle),
  };
}

File: src/execution/host/http.rs
//! HTTP host functions (the `Dream` module behind `system.net` `HttpClient`). The buffered
//! `httpRequest*` functions perform the whole request synchronously (blocking `reqwest`) and
//! bridge the serialized response into Dream's async runtime. The streaming `httpRequestStream*`
//! functions keep the `reqwest::blocking::Response` in a handle table so `httpReadChunk` can
//! pull the body incrementally without buffering it all in memory.

use std::io::Read;
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::{Mutex, OnceLock};
use std::time::Duration;

use indexmap::IndexMap;

fn build_http_client(http2: bool) -> reqwest::blocking::Client {
    let mut builder = reqwest::blocking::Client::builder()
        .pool_max_idle_per_host(0)
        .pool_idle_timeout(Duration::from_secs(0))
        .tcp_keepalive(None)
        .connect_timeout(Duration::from_secs(10));
    if !http2 {
        builder = builder.http1_only();
    }
    builder.build().expect("reqwest HTTP client")
}

/// Fresh client per request so idle connection state is not reused across calls.
fn http_client(http_version: i32) -> reqwest::blocking::Client {
    build_http_client(http_version == 2)
}

fn http_transport_error(msg: &str) -> Vec<u8> {
    let mut out = b"0\n\n".to_vec(); // status 0 = transport error; body is the message
    out.extend_from_slice(msg.as_bytes());
    out
}

fn http_error_message(err: &reqwest::Error) -> String {
    if err.is_timeout() {
        "timeout".to_string()
    } else {
        err.to_string()
    }
}

/// Runs `f` with an optional wall-clock budget. When `wall` is `None`, runs on the calling thread.
/// When set, runs on a worker and returns `None` if the budget elapses before `f` finishes; the
/// worker result is then discarded (never published to the caller).
fn with_http_wall<T, F>(wall: Option<Duration>, f: F) -> Option<T>
where
    T: Send + 'static,
    F: FnOnce() -> T + Send + 'static,
{
    let Some(wall) = wall else {
        return Some(f());
    };
    let (tx, rx) = std::sync::mpsc::channel();
    std::thread::spawn(move || {
        let _ = tx.send(f());
    });
    rx.recv_timeout(wall).ok()
}

fn response_head(response: &reqwest::blocking::Response) -> String {
    let status = response.status().as_u16();
    let mut head = format!("{}\n", status);
    for (name, value) in response.headers().iter() {
        if let Ok(v) = value.to_str() {
            head.push_str(name.as_str());
            head.push_str(": ");
            head.push_str(v);
            head.push('\n');
        }
    }
    head.push('\n');
    head
}

/// Performs one blocking HTTP request and serializes the whole response into the wire format shared
/// with `runtime/dream.js` (and parsed by `HttpResponse`): an ASCII head ("<status>\n" plus
/// "Name: value\n" lines), a blank line, then the raw body bytes. `body` is sent verbatim unless the
/// verb is GET/HEAD or it is empty. Network/protocol errors come back as a status `0` response whose
/// body is the error text. `timeout_ms == 0` means no timeout. `http_version` is `1` (default) or `2`.
pub(crate) fn perform_http(
    method: &str,
    url: &str,
    headers_json: &str,
    body: Vec<u8>,
    timeout_ms: i32,
    http_version: i32,
) -> Vec<u8> {
    let method = method.to_owned();
    let url = url.to_owned();
    let headers_json = headers_json.to_owned();
    let wall = if timeout_ms > 0 {
        Some(Duration::from_millis(timeout_ms as u64))
    } else {
        None
    };
    match with_http_wall(wall, move || {
        match build_request(&method, &url, &headers_json, body, timeout_ms, http_version).send() {
            Ok(response) => {
                let mut out = response_head(&response).into_bytes();
                if let Ok(body_bytes) = response.bytes() {
                    out.extend_from_slice(&body_bytes);
                }
                out
            }
            Err(e) => http_transport_error(&http_error_message(&e)),
        }
    }) {
        Some(out) => out,
        None => http_transport_error("timeout"),
    }
}

/// Builds a `reqwest::blocking::RequestBuilder` shared by the buffered and streaming request
/// paths, applying method/headers/timeout/body identically.
fn build_request(
    method: &str,
    url: &str,
    headers_json: &str,
    body: Vec<u8>,
    timeout_ms: i32,
    http_version: i32,
) -> reqwest::blocking::RequestBuilder {
    let verb = method.to_uppercase();
    let http_method = reqwest::Method::from_bytes(verb.as_bytes()).unwrap_or(reqwest::Method::GET);

    let mut builder = http_client(http_version).request(http_method, url);
    if timeout_ms > 0 {
        builder = builder.timeout(Duration::from_millis(timeout_ms as u64));
    }

    if !headers_json.is_empty() {
        if let Ok(serde_json::Value::Object(map)) =
            serde_json::from_str::<serde_json::Value>(headers_json)
        {
            for (name, value) in map.iter() {
                if let Some(v) = value.as_str() {
                    builder = builder.header(name.as_str(), v);
                }
            }
        }
    }

    if !body.is_empty() && verb != "GET" && verb != "HEAD" {
        builder = builder.body(body);
    }
    builder
}

fn stream_handles() -> &'static Mutex<IndexMap<u32, reqwest::blocking::Response>> {
    static HANDLES: OnceLock<Mutex<IndexMap<u32, reqwest::blocking::Response>>> = OnceLock::new();
    HANDLES.get_or_init(|| Mutex::new(IndexMap::new()))
}

static NEXT_STREAM_HANDLE: AtomicU32 = AtomicU32::new(1);

/// Opens a request and returns just the head (status + headers), keeping the response body
/// unread in a handle table for `http_read_chunk` to pull incrementally. Wire format matches
/// [`perform_http`] exactly except the "body" is the decimal handle id, so `HttpResponse`'s
/// existing head parser can be reused unchanged on the Dream side (see `wrap_stream` in
/// `http_client.dream`).
pub(crate) fn open_http_stream(
    method: &str,
    url: &str,
    headers_json: &str,
    body: Vec<u8>,
    timeout_ms: i32,
    http_version: i32,
) -> Vec<u8> {
    let method = method.to_owned();
    let url = url.to_owned();
    let headers_json = headers_json.to_owned();
    let wall = if timeout_ms > 0 {
        Some(Duration::from_millis(timeout_ms as u64))
    } else {
        None
    };
    // Send the Response back to this thread before publishing a stream handle, so a wall timeout
    // cannot leave an orphan entry in the handle table.
    match with_http_wall(wall, move || {
        build_request(&method, &url, &headers_json, body, timeout_ms, http_version)
            .send()
            .map_err(|e| http_error_message(&e))
    }) {
        None => http_transport_error("timeout"),
        Some(Err(msg)) => http_transport_error(&msg),
        Some(Ok(response)) => {
            let mut head = response_head(&response);
            let id = NEXT_STREAM_HANDLE.fetch_add(1, Ordering::Relaxed);
            let mut table = stream_handles().lock().unwrap_or_else(|e| e.into_inner());
            table.insert(id, response);
            head.push_str(&id.to_string());
            head.into_bytes()
        }
    }
}

/// Reads up to `max_bytes` from an open stream handle. Wire: `"data\n<bytes>"` | `"eof\n"` |
/// `"error\n<message>"`.
pub(crate) fn http_read_chunk(handle: i32, max_bytes: i32) -> Vec<u8> {
    let mut table = stream_handles().lock().unwrap_or_else(|e| e.into_inner());
    let Some(response) = table.get_mut(&(handle as u32)) else {
        return b"error\nstream not found".to_vec();
    };
    let cap = (max_bytes.max(1) as usize).min(1 << 20);
    let mut buf = vec![0u8; cap];
    match response.read(&mut buf) {
        Ok(0) => {
            drop(table);
            let mut table = stream_handles().lock().unwrap_or_else(|e| e.into_inner());
            table.shift_remove(&(handle as u32));
            b"eof\n".to_vec()
        }
        Ok(n) => {
            let mut out = b"data\n".to_vec();
            out.extend_from_slice(&buf[..n]);
            out
        }
        Err(e) => {
            let mut out = b"error\n".to_vec();
            out.extend_from_slice(e.to_string().as_bytes());
            out
        }
    }
}

pub(crate) fn http_close_stream(handle: i32) -> i32 {
    let mut table = stream_handles().lock().unwrap_or_else(|e| e.into_inner());
    table.shift_remove(&(handle as u32)).is_some() as i32
}
